### PR TITLE
Bringing back Watching Folder feature

### PR DESCRIPTION
**Pull request by numericalatina**
[This is still a work-in-process]

I'm opening the pull request to let the community know I'm working on this feature, in case there is someone else making the same effort, so we can concurr.

### CODE REVIEW

**Comment by numericalatina** [comment]
This work was based upon #111

**Comment by paddatrapper** [comment]
I've added [WIP] to the title to indicate that it is still a Work in Progress. The bot ensures that we do not accidentally merge this before it is ready. When you want a review, edit the title, remove the [WIP] and ping me/one of the other maintainers

**Comment by numericalatina** [comment]
I have pushed the fixes the came out of testing this scripts against > 60,000 audios (over a Tb)
This is working for my needs, and takes aprox. 40min to scan the whole of it, so I programmed it hourly in the crontab.
Of course it could be still cleaned up a little bit more, in syntax and may be DRY some stanzas (such as the MIME formats, for ex.). But I'm removing the WIP tag so that you could please give it a review.

File: python_apps/libretime_watch/libretime_watch/libretime_watch.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+def connect_to_messaging_server():
+  """Connect to RabbitMQ Server and start listening for messages. 
+     Returns RabbitMQ connection and channel
+  """
+  credentials=pika.credentials.PlainCredentials(config["rm_user"],config["rm_pass"])
+  connection = pika.BlockingConnection(pika.ConnectionParameters(host=config["rm_host"],
+            virtual_host=config["rm_vhost"],credentials=credentials))
+  channel = connection.channel()
+#  channel.exchange_delete (exchange=EXCHANGE)
```
Comment (paddatrapper) [inline python_apps/libretime_watch/libretime_watch/libretime_watch.py]: Please delete commented out code

File: python_apps/libretime_watch/libretime_watch/start_watching.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+# initialize logging
+logfile= "/var/log/airtime/libretime_watch_cron.log"
+logging.basicConfig(format='%(asctime)s %(message)s',filename=logfile,level=logging.INFO)
+
+EXCHANGE="airtime-media-monitor"
+EXCHANGE_TYPE = "direct"
+ROUTING_KEY="filesystem"
+QUEUE="media-monitor"
+
+CONFIGFILE="/etc/airtime/airtime.conf"
```
Comment (paddatrapper) [inline python_apps/libretime_watch/libretime_watch/start_watching.py]: Can't this be imported from `readconfig`? That way you don't have to re-read all the config from the disk. You would just need to add the RabbitMQ stuff to `readconfig`

File: python_apps/libretime_watch/setup.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+from subprocess import call
+import sys
+import os
+
+# Change directory since setuptools uses relative paths
+script_path = os.path.dirname(os.path.realpath(__file__))
+print script_path
+os.chdir(script_path)
+
+# Allows us to avoid installing the upstart init script when deploying airtime_analyzer
```
Comment (paddatrapper) [inline python_apps/libretime_watch/setup.py]: This comment is out of date

File: python_apps/libretime_watch/setup.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      install_requires=[
+          'mutagen',
+          'pika',
+          'psycopg2', # database
+          'jason'
+      ],
+      zip_safe=False,
+      data_files=data_files)
+
+# Remind users to reload the initctl config so that "service start airtime_analyzer" works
```
Comment (paddatrapper) [inline python_apps/libretime_watch/setup.py]: airtime_analyzer -> libretime_watch

File: python_apps/libretime_watch/test/test_rabbit.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+   cur.execute ("SELECT id,directory from cc_music_dirs where type = 'watched'")
+   row = cur.fetchone()
+   id = row[0]
+   watch_dir = row[1]
+   cur.close()
+except:
+   print ("Can't get directory for watching")
+   exit()
+
+#print ("ID: "+str(id)+" Dir: "+str(watch_dir)+ " "+config['api_key'])
```
Comment (paddatrapper) [inline python_apps/libretime_watch/test/test_rabbit.py]: Please delete

File: python_apps/libretime_watch/test/test_rabbit.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/python
```
Comment (paddatrapper) [inline python_apps/libretime_watch/test/test_rabbit.py]: What is thus actually testing?

**Comment by paddatrapper** [comment]
The other thing is please add docs on how to use this feature. See http://libretime.org/manual/media-folders/

**Comment by Robbt** [comment]
So I finally got around to testing this and I haven't been able to get it to work. Starting the service appears to happen but then it fails to do anything. I can start and stop it but I don't see a libretime_watch process. So trying to manually start it from the /usr/local/lib/python2.7/dist-packages/libretime_watch-0.1-py2.7.egg/libretime_watch directory I get this error.
'''
Traceback (most recent call last):
  File "libretime_watch.py", line 259, in <module>
    main()
  File "libretime_watch.py", line 240, in main
    connection, channel = connect_to_messaging_server()
  File "libretime_watch.py", line 181, in connect_to_messaging_server
    credentials=pika.credentials.PlainCredentials(config["rm_user"],config["rm_pass"])
KeyError: 'rm_user'
'''

**Comment by kmahelona** [comment]
Well I had a bunch written here then the laptop crashed...

Anyway I've been trying this out and have experienced the same issues mentioned above by @Robbt. 

One interesting thing I finally found out is that the watch_folder stuff has a limited amount of metadata that's imported compare with the [metadata_analyzer.py](https://github.com/LibreTime/libretime/blob/5b822afe14046157a8d9fbdff103eb383673e6be/python_apps/airtime_analyzer/airtime_analyzer/metadata_analyzer.py). For years I've been unable to import the `label` field via `watch_folders` and it's because the [metadata.py](https://github.com/althost/libretime/blob/libretime_watch_rebased/python_apps/libretime_watch/libretime_watch/metadata.py) script just doesn't import all possible options.

Anyway I think we should mirror what's done in `metadata_analyzer.py`.

I'm happy to work on some updates to this PR since this is a show stopper for us getting LibreTime working in production. The `type` to `exchange_type` changes are pretty trivial.

The metadata stuff looks a little tricker because there's some hardcoded character length limits for the database which I don't know what they should be e.g. for `label` but I feel like we should just import `metadata_analyzer.py` and use that?

I think this feature is preferred for us over the `libretime_import` feature.

**Comment by kmahelona** [comment]
An update on this.

**Setup**
- Ubuntu 18.04.4 LTS.
- Modified field names for `start_watching.py` to get the script to work.
- I made a few modifications to `metadata_analyzer.py` - I've added the `label` field
- I made a cron job that runs `python /usr/local/lib/python2.7/dist-packages/libretime_watch-0.1-py2.7.egg/libretime_watch/start_watching.py` every minute.

**Observations**
- I've noticed that while the `libretime_watch` service was still registered as running with no errors showing from `service libretime_watch status`, no new files were being scanned. I had to manually restart the service to get it working again. Any help to investigate the cause of this would be appreciated. The only errors I see in the `libretime_watch.log` are related to rabbitmq connections but scans seemed to happen after these errors.
- The watch field does recurse into sub directories. I like this, but the docs should mention this. Also should we ignore certain folders e.g. ones that begin with `.`? In my case since I have a Resilio Sync folder, there are `.sync` folders which I'd want to ignore.
- **[BUG]** Updates to file metadata aren't imported/synced with the database. Airtime's implementation would sync metadata changes, and I think this an important aspect of the watch_folder functionality.

**Comment by kmahelona** [comment]
Hi,

I've forked this branch here: https://github.com/TeHikuMedia/libretime/tree/libretime_watch_rebased

I've made a number of changes to get this to work. More recent ones include:
- Fixed the cron job (though I've hardcoded the path to the python file which is probably bad?)
- Supports multiple watch folders

I'd be keen to see this PR merged into master but that would have to come from @numericalatina right? Not sure how I can merge my changes into this PR.

Thx

**Comment by paddatrapper** [comment]
> I'd be keen to see this PR merged into master but that would have to come from @numericalatina right? Not sure how I can merge my changes into this PR.

You can send patches using `git format-patch` as attachments to comments on this PR or submit a PR to althost/libretime against the libretime_watch_rebased branch. We should be able to review and merge changes made there if I understand the modification permissions granted by GitHub on PRs correctly.

**Comment by paddatrapper** [comment]
The solution to the cron job is probably to get the python package to install a CLI entry point that is added to the PATH. See [here](https://setuptools.readthedocs.io/en/latest/setuptools.html#automatic-script-creation) for the explanation

**Comment by kmahelona** [comment]
Thx @paddatrapper. I just submitted a PR. Still new to this side of GitHub but it's awesome! I'll work on the cron job issue later (it's getting late in Aotearoa).

**Comment by kmahelona** [comment]
Currently the `Delete All Tracks` appears to delete things in watch folders. So until that's fixed let's definitely not merge this.

**Comment by kmahelona** [comment]
### Should we delete files in the watched directory if they are deleted from Libretime?
My initial answer to this was *yes* but now that I've started to implement this I think that's the wrong answer. If we don't delete the physical file, then it will get automatically imported on a rescan. To prevent that, we'd have to keep some information in the database or somewhere else. That probably complicates things too much. So maybe the right answer is *no* and we have a warning in the docs that says files put in the `watch_folder` can be deleted.

What are your thoughts about this?

**Comment by paddatrapper** [comment]
Definitely the simplest option would be to delete the file. I think we should probably do as you say and just note that files may be deleted from the watch directory if they are removed from LibreTime. If we find that there is a significant use case for keeping the files if they are deleted (tracked by issue requests and forum posts), we can look at addition that functionality.

For now, I think simplicity is the best way forward

**Comment by kmahelona** [comment]
I'd like to implement importing of album artwork with this, but it's not required for a working watching folder. I'm not sure how exactly to do this. I can extract the images from files and save them to various sizes but I don't know how to tell LibreTime that the artwork is there. I've set something like `database['artwork']=/path/to/artwork_folder` to the `ccFiles` table but that doesn't seem to do anything.

**Comment by Robbt** [comment]
Honestly as far as deleting files go from LibreTime I think the ideal solution would be to move the files to a deleted files section that can be manually purged and also to retain the metadata in the database but mark it as "deleted" or "inactive". This I guess is a separate concern from "watched folders" but would be one way of preventing catastrophic data loss by someone who was running LibreTime and didn't realize that deleting items from the web UI would affect folders they watched. 

Or another option would be to create a deleted or inactive section of the library and when someone deletes a file that is a "watched folder" it wouldn't automatically reimport because it already exists in the database albeit marked as "deleted". 

I think there are benefits here in terms of retaining metadata etc and preventing accidental deletion of important files by users trying to clean up the library. In my opinion physically deleting a file should be at least a two step process with validation. But that is just my opinion, we have had program managers delete audio files that other DJs were then annoyed that they were missing.

I guess my solution is massively more complicated though so I'm not going to suggest we implement it as an alternative without the time to actually code it. I really don't want someone to accidentally delete their parts or their entire music collection while testing LibreTime out though.

**Comment by kmahelona** [comment]
I agree on a two step process for deleting. I've destroyed our entire database twice. While I knew what I was doing I forgot that I had an IP address in an airtime.conf file that I didn't update so when I was working on a "test" machine it was actually doing API calls to my production machine. Maybe it's a feature we can work on.

**Comment by kmahelona** [comment]
A few more notes on things to check/do:

- Update airtime INSTALL to include this service
  - ensure this is done correctly. currently if libretime_watch dies `service libretime_watch status` still shows it's running.
  - sometimes stopping the libretime_watch service doesn't actually work and you have to do a `sudo kill -9 1234`
- Cody tidying

**Comment by paddatrapper** [comment]
Also need to update the docs mentioning `systemctl restart`, etc

**Comment by kmahelona** [comment]
We need to address #845 before this PR is ready.

**Comment by geedos** [comment]
> I've forked this branch here: https://github.com/TeHikuMedia/libretime/tree/libretime_watch_rebased

Hey guys,

I've tried to install a bunch of the different branches with the watch feature (including the one quoted above), but all of the branches I've tried are failing on the install with :

`Exception: Python 2 no longer supported`

This doesn't happen when I clone and install the master, it installs fine.

Can you advise what I need to update to get this branch working on Ubuntu 18.04.4 LTS please?

It seems to bomb out here if it helps :

`File "/tmp/easy_install-7f1Bvn/mutagen-1.44.0/setup.py", line 246, in <module>`

Many thanks!

**Comment by kmahelona** [comment]
Kia ora. This feature only works with python3 as I believe we’re dropping python2 support. I think the right approach would be to go off the master branch and then merge this branch locally. I’m sorry I haven’t merged master into this branch (our fork you quoted) in a while (For various reason) but that shouldn’t be necessary there are no merge conflicts. I’d like to get this merged into master but I’ve been super busy and there’s a few things to check off the list before that happens.

**Comment by geedos** [comment]
Thanks for the speedy reply.

I tried the solutions offered here : [https://discourse.libretime.org/t/exception-python-2-no-longer-supported/466](url) before realising that, so of course this didn't work either.

I've gone back to the master now so everything is working again but I'm no git guru sadly. Can you advise how I would merge this branch locally? Sorry if it's a really nooby question.

**Comment by kmahelona** [comment]
All good. I was editing my answer then my browser refreshed. So the problem is that mutagen I’m pretty sure (I had this problem before). 1.44 drops python2 support and being on Ubuntu 18 where the default is python2 the installer is probably using python2 and trying to install 1.44. 

Anyway, yea I’m not a git guru either lol but there is a way to do this as I’ve done it after some googling. https://stackoverflow.com/questions/9153598/how-do-i-fetch-a-branch-on-someone-elses-fork-on-github#9153737 might do the trick. 

Alternatively, you could change this line so that it looks exactly like what’s here: https://github.com/TeHikuMedia/libretime/blob/931933b55b3ef4cd2070553c10b45b919f310e48/python_apps/libretime_watch/setup.py#L45

**Comment by paddatrapper** [comment]
Python 2 support has been dropped since #951 has been merged

**Comment by xabispacebiker** [comment]
Hi mates, I am putting my hands on this too.. I havent read all comments and/or related documentation yet but from what I've read so far, I ask myself why we are not using python watchdog to check for modified files instead of rescanning / importing the whole directories ? Please, excuse me if I am saying something that you have already talked about but then I have not reached that section yet.

**Comment by kmahelona** [comment]
Hey @xabispacebiker that sounds like a better approach. I think much of this code was copied over from the old Airtime just to get it re-implemented in LibreTime asap. In terms of getting this PR over the line, the last thing I think we were needing is to ensure the new "types" UI was implemented for watch folders.

Here's the relevant PR for this PR: https://github.com/althost/libretime/pull/4.

**Comment by xabispacebiker** [comment]
soon i will start working on this, so you guys think is a good idea to switch to python's watchdog ? It does not seem to difficult to implement 

http://thepythoncorner.com/dev/how-to-create-a-watchdog-in-python-to-look-for-filesystem-changes/

**Comment by diroots** [comment]
hi there, about a watchdog for filesystem changes, iwatch tool has interesting features, as it can detect events like the end of file write, file deletion,... 

http://iwatch.sourceforge.net/documentation.html

**Comment by zklosko** [comment]
@numericalatina Can you please force-push this branch to be level with the updates made to the master branch? This way we can start testing again.

**Comment by numericalatina** [comment]
Sorry but I will not force-push this branch, I have to put it up to date with master, which will probably bring new issues, as has been the case before... i consider it is a lot of work to update my version of LibreTime.
I hope i will have time to do it next year.
Cheers

**Comment by kmahelona** [comment]
Hey @zklosko I've had the same caution as @numericalatina. Our production radio station is running as we need it and merging current master would mean possible issues and downtime with our broadcast. I'm not sure when I'll have time, hopefully in the next couple months (maybe sooner as we're requiring that non-flash export playout history) I could merge master into PR https://github.com/althost/libretime/pull/4 but we'd still be waiting on this PR to update with master.

**Comment by jfloury** [comment]
Hi everybody,

Is this PR going on stage ? We are looking forward to this functionality !
Is there a way to watch folders ?

Tanks

**Comment by paddatrapper** [comment]
This PR had currently stalled waiting on someone to finish it. Currently watch folders are not supported

**Comment by jfloury** [comment]
> This PR had currently stalled waiting on someone to finish it. Currently watch folders are not supported

Thanks.

I don't know why de devs are on another repo by kmahelona ? why isn't it in a branch of current repo ?

**Comment by paddatrapper** [comment]
Because we follow C4 development guidelines - not everyone usd push access to this repo. PRs are made by committing changes to your fork and opening a PR to merge it into master of this project

**Pull request closed without merging**